### PR TITLE
Allow using any polkadot client instead of enum Client

### PR DESCRIPTION
**Pull request by cecton**
Related to https://github.com/paritytech/cumulus/issues/199

### CODE REVIEW

File: collator/src/lib.rs [hunk 1]
```diff
@@ -67,16 +67,17 @@ pub use sc_network::PeerId;
...
+use polkadot_service::{FullNodeHandles, AbstractClient, YaExecuteWithClient};
 #[cfg(feature = "service-rewr")]
 use polkadot_service_new::{
 	self as polkadot_service,
 	Error as ServiceError, FullNodeHandles, AbstractClient,
 };
 use sc_service::SpawnTaskHandle;
 use sp_core::traits::SpawnNamed;
-use sp_runtime::traits::BlakeTwo256;
+use sp_runtime::traits::{Block as BlockT, BlakeTwo256};
```
Comment (cecton) [inline collator/src/lib.rs:78]:
```suggestion
use sp_runtime::traits::BlakeTwo256;
```

File: service/src/client.rs [hunk 2]
```diff
@@ -78,6 +78,8 @@ pub trait AbstractClient<Block, Backend>:
 		Self::Api: RuntimeApiCollection<StateBackend = Backend::State>,
 {}
 
+//impl<Block, Backend, Client> AbstractClient<Block, Backend> for Arc<Client> {}
+
```
Comment (cecton) [inline service/src/client.rs:82]:
```suggestion
```

File: service/src/client.rs [hunk 3]
```diff
@@ -118,6 +118,12 @@ pub trait ExecuteWithClient {
 			Client: AbstractClient<Block, Backend, Api = Api> + 'static;
 }
 
+/// Yet Another ExecuteWithClient
+pub trait YaExecuteWithClient {
+	/// Execute the given something with the client.
+	fn execute_with<T: ExecuteWithClient>(&self, t: T) -> T::Output;
+}
```
Comment (cecton) [inline service/src/client.rs:125]: I'm not sure how to call it... ClientExecutor? ClientExecutionContext? @bkchr

File: service/src/client.rs [hunk 4]
```diff
@@ -118,6 +118,12 @@ pub trait ExecuteWithClient {
 			Client: AbstractClient<Block, Backend, Api = Api> + 'static;
 }
 
+/// Yet Another ExecuteWithClient
+pub trait YaExecuteWithClient {
```
Comment (bkchr) [inline service/src/client.rs:122]: Maybe `ClientHandle`?

With a much better documentation on what it does and why we need it? Something like:
```
/// A handle to a Polkadot client instance
///
/// The Polkadot service supports multiple different runtimes, like
/// Westend, Polkadot itself, etc. As each runtime has a specialized
/// client, we need to hide them behind a trait. This is this trait. 
///
/// When wanting to work with the inner client, you need to use [`execute_with`](ClientHandle::execute_with).
/// See [`ExecuteWithClient`] for more information.
```

File: collator/src/lib.rs [hunk 5]
```diff
@@ -117,14 +118,19 @@ pub trait BuildParachainContext {
...
 		self,
-		client: polkadot_service::Client,
+		client: Arc<Client>,
 		spawner: SP,
 		network: impl Network + SyncOracle + Clone + 'static,
 	) -> Result<Self::ParachainContext, ()>
 		where
-			SP: SpawnNamed + Clone + Send + Sync + 'static;
+			SP: SpawnNamed + Clone + Send + Sync + 'static,
+			Backend: BackendT<Block>,
```
[the remark this answers is not in the document]
Comment (cecton) [reply collator/src/lib.rs:129]: I AM SO SORRY

File: service/src/client.rs (hunk 4 above)
Comment (cecton) [reply service/src/client.rs:122]: omg I was not expecting that much 😆 :pray:

**Comment by cecton** [comment]
bot merge

**Comment by cecton** [comment]
bot merge

**Pull request merged**